README
README file for Javascript::SHA1.

Warning: WinZip 8.1 and 9.0 both contain an 'accidental' bug which stops
them recognizing the directory structures of some valid *.tar files.
You are better off using a reliable tool such as InfoZip:
ftp://ftp.info-zip.org/pub/infozip/
This bug has something to do with POSIX style paths.

1 Installing from a Unix-like distro
------------------------------------
shell>gunzip Javascript-SHA1-1.00.tgz
shell>tar mxvf Javascript-SHA1-1.00.tar

On Unix-like systems:

shell>perl Build.PL
shell>./Build
shell>./Build test
shell>./Build install

On MS Windows-like systems:

shell>perl Build.PL
shell>perl Build
shell>perl Build test
shell>perl Build install

On all systems:

Run SHA1.pm through you favourite pod2html translator.

If you are using fancy-pom2.pl, with its 'default.css' file installed in
/apache2/htdocs/css/, you'd do:

	shell>perl fancy-pom2.pl html -css SHA1.pm > /apache2/htdocs/SHA1.html

	or perhaps something like:

	shell>perl fancy-pom2.pl html -css SHA1.pm > /perl/html/site/lib/Javascript/SHA1.html

2 Installing from an ActiveState distro
---------------------------------------
shell>unzip Javascript-SHA1-1.00.zip
shell>ppm install --location=. Javascript-SHA1
shell>del Javascript-SHA1-1.00.ppd
shell>del PPM-Javascript-SHA1-1.00.tar.gz

Changes.txt
Revision history for Perl extension Javascript::SHA1.

1.04  Thu Oct 23 18:11:00 2008
	- Remove ^M chars from source, including within the Javascript

1.03  Sat Aug 04 13:52:00 2007
	- Extend method javascript() to take a second parameter, which is the
		number of the CGI form within the HTML page. The allows the module to
		operate on an HTML page containing more than one CGI form.
		The default value is 0, which is the first form, and perhaps the only form,
		on the page. So, no need to upgrade if you only have one form per page
	- Update the docs to describe this new parameter
	- Note: Original patch kindly supplied by Thomas Goik
	
1.02  Fri Oct 21 14:27:00 2005
	- No code changes (ouside new() ), so no need to upgrade
	- Simplify code in new() to no longer check if caller is an object
		as per the new standard on such matters. Personally, I don't care :-)
	- Add more things to MANIFEST.SKIP
	- Regenerate META.yml with Module::Build V 0.2611 to correct a faulty
		META.yml output by V 0.24 which falsely said 'provides: {}',
		which in turn stopped the PAUSE indexer from finding the SHA1.pm file,
		which in turn meant this module was omitted from the master index:
		ftp://pause.perl.org/pub/PAUSE/modules/01modules.index.html

1.01  Tue Jul 20 20:43:00 2004
	- Change Makefile.PL to coexist with Module::Build
	- Add t/pod.t to test all PODs

1.00  Fri Mar  05 10:23:29 2004
	- Original version
	- The Javascript is Paul Johnston's version, from his web site:
		http://pajhome.org.uk/crypt/md5
	- There are 3 functions, str2hex_sha1(), str2b64_sha1() and str2str_sha1(), to return various versions
		of the digest


lib/Javascript/SHA1.pm
package Javascript::SHA1;

# Name:
#	Javascript::SHA1.
#
# Documentation:
#	POD-style documentation is at the end. Extract it with pod2html.*.
#
# Reference:
#	Object Oriented Perl
#	Damian Conway
#	Manning
#	1-884777-79-1
#	P 114
#
# Note:
#	o Tab = 4 spaces || die.
#
# Author:
#	Ron Savage <ron@savage.net.au>
#	Home page: http://savage.net.au/index.html
#
# Licence:
#	Australian copyright (c) 2004 Ron Savage.
#
#	All Programs of mine are 'OSI Certified Open Source Software';
#	you can redistribute them and/or modify them under the terms of
#	The Artistic License, a copy of which is available at:
#	http://www.opensource.org/licenses/index.html

use strict;
use warnings;
no warnings 'redefine';

require 5.005_62;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Javascript::SHA1 ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(

) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(

);
our $VERSION = '1.04';

# -----------------------------------------------

# Preloaded methods go here.

# -----------------------------------------------

# Encapsulated class data.

{
	my(%_attr_data) =
	(	# Alphabetical order.
	);

	sub _default_for
	{
		my($self, $attr_name) = @_;

		$_attr_data{$attr_name};
	}

	sub _standard_keys
	{
		keys %_attr_data;
	}

}	# End of encapsulated class data.

# -----------------------------------------------

sub javascript
{
	my($self, $field_name, $form_number) = @_;
	$field_name		||= 'password';
	$form_number	||= 0;

	return <<EOS;
/*
 * A JavaScript implementation of the Secure Hash Algorithm, SHA-1, as defined
 * in FIPS PUB 180-1
 * Version 2.1 Copyright Paul Johnston 2000 - 2002.
 * Other contributors: Greg Holt, Andrew Kepert, Ydnar, Lostinet
 * Distributed under the BSD License
 * See http://pajhome.org.uk/crypt/md5 for details.
 * Patch by Ron Savage 2004-04-27:
 * o Add the 3 functions str2hex_sha1, str2b64_sha1 and str2str_sha1().
 */

function str2hex_sha1()
{
	document.forms[$form_number].$field_name.value = hex_sha1(document.forms[$form_number].$field_name.value);
	return true;
}

function str2b64_sha1()
{
	document.forms[$form_number].$field_name.value = b64_sha1(document.forms[$form_number].$field_name.value);
	return true;
}

function str2str_sha1()
{
	document.forms[$form_number].$field_name.value = str_sha1(document.forms[$form_number].$field_name.value);
	return true;
}

/*
 * Configurable variables. You may need to tweak these to be compatible with
 * the server-side, but the defaults work in most cases.
 */
var hexcase = 0;  /* hex output format. 0 - lowercase; 1 - uppercase        */
var b64pad  = ""; /* base-64 pad character. "=" for strict RFC compliance   */
var chrsz   = 8;  /* bits per input character. 8 - ASCII; 16 - Unicode      */

/*
 * These are the functions you'll usually want to call.
 * They take string arguments and return either hex or base-64 encoded strings
 */
function hex_sha1(s){return binb2hex(core_sha1(str2binb(s),s.length * chrsz));}
function b64_sha1(s){return binb2b64(core_sha1(str2binb(s),s.length * chrsz));}
function str_sha1(s){return binb2str(core_sha1(str2binb(s),s.length * chrsz));}
function hex_hmac_sha1(key, data){ return binb2hex(core_hmac_sha1(key, data));}
function b64_hmac_sha1(key, data){ return binb2b64(core_hmac_sha1(key, data));}
function str_hmac_sha1(key, data){ return binb2str(core_hmac_sha1(key, data));}

/*
 * Perform a simple self-test to see if the VM is working
 */
function sha1_vm_test()
{
  return hex_sha1("abc") == "a9993e364706816aba3e25717850c26c9cd0d89d";
}

/*
 * Calculate the SHA-1 of an array of big-endian words, and a bit length
 */
function core_sha1(x, len)
{
  /* append padding */
  x[len >> 5] |= 0x80 << (24 - len % 32);
  x[((len + 64 >> 9) << 4) + 15] = len;

  var w = Array(80);
  var a =  1732584193;
  var b = -271733879;
  var c = -1732584194;
  var d =  271733878;
  var e = -1009589776;

  for(var i = 0; i < x.length; i += 16)
  {
    var olda = a;
    var oldb = b;
    var oldc = c;
    var oldd = d;
    var olde = e;

    for(var j = 0; j < 80; j++)
    {
      if(j < 16) w[j] = x[i + j];
      else w[j] = rol(w[j-3] ^ w[j-8] ^ w[j-14] ^ w[j-16], 1);
      var t = safe_add(safe_add(rol(a, 5), sha1_ft(j, b, c, d)),
                       safe_add(safe_add(e, w[j]), sha1_kt(j)));
      e = d;
      d = c;
      c = rol(b, 30);
      b = a;
      a = t;
    }

    a = safe_add(a, olda);
    b = safe_add(b, oldb);
    c = safe_add(c, oldc);
    d = safe_add(d, oldd);
    e = safe_add(e, olde);
  }
  return Array(a, b, c, d, e);

}

/*
 * Perform the appropriate triplet combination function for the current
 * iteration
 */
function sha1_ft(t, b, c, d)
{
  if(t < 20) return (b & c) | ((~b) & d);
  if(t < 40) return b ^ c ^ d;
  if(t < 60) return (b & c) | (b & d) | (c & d);
  return b ^ c ^ d;
}

/*
 * Determine the appropriate additive constant for the current iteration
 */
function sha1_kt(t)
{
  return (t < 20) ?  1518500249 : (t < 40) ?  1859775393 :
         (t < 60) ? -1894007588 : -899497514;
}

/*
 * Calculate the HMAC-SHA1 of a key and some data
 */
function core_hmac_sha1(key, data)
{
  var bkey = str2binb(key);
  if(bkey.length > 16) bkey = core_sha1(bkey, key.length * chrsz);

  var ipad = Array(16), opad = Array(16);
  for(var i = 0; i < 16; i++)
  {
    ipad[i] = bkey[i] ^ 0x36363636;
    opad[i] = bkey[i] ^ 0x5C5C5C5C;
  }

  var hash = core_sha1(ipad.concat(str2binb(data)), 512 + data.length * chrsz);
  return core_sha1(opad.concat(hash), 512 + 160);
}

/*
 * Add integers, wrapping at 2^32. This uses 16-bit operations internally
 * to work around bugs in some JS interpreters.
 */
function safe_add(x, y)
{
  var lsw = (x & 0xFFFF) + (y & 0xFFFF);
  var msw = (x >> 16) + (y >> 16) + (lsw >> 16);
  return (msw << 16) | (lsw & 0xFFFF);
}

/*
 * Bitwise rotate a 32-bit number to the left.
 */
function rol(num, cnt)
{
  return (num << cnt) | (num >>> (32 - cnt));
}

/*
 * Convert an 8-bit or 16-bit string to an array of big-endian words
 * In 8-bit function, characters >255 have their hi-byte silently ignored.
 */
function str2binb(str)
{
  var bin = Array();
  var mask = (1 << chrsz) - 1;
  for(var i = 0; i < str.length * chrsz; i += chrsz)
    bin[i>>5] |= (str.charCodeAt(i / chrsz) & mask) << (24 - i%32);
  return bin;
}

/*
 * Convert an array of big-endian words to a string
 */
function binb2str(bin)
{
  var str = "";
  var mask = (1 << chrsz) - 1;
  for(var i = 0; i < bin.length * 32; i += chrsz)
    str += String.fromCharCode((bin[i>>5] >>> (24 - i%32)) & mask);
  return str;
}

/*
 * Convert an array of big-endian words to a hex string.
 */
function binb2hex(binarray)
{
  var hex_tab = hexcase ? "0123456789ABCDEF" : "0123456789abcdef";
  var str = "";
  for(var i = 0; i < binarray.length * 4; i++)
  {
    str += hex_tab.charAt((binarray[i>>2] >> ((3 - i%4)*8+4)) & 0xF) +
           hex_tab.charAt((binarray[i>>2] >> ((3 - i%4)*8  )) & 0xF);
  }
  return str;
}

/*
 * Convert an array of big-endian words to a base-64 string
 */
function binb2b64(binarray)
{
  var tab = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
  var str = "";
  for(var i = 0; i < binarray.length * 4; i += 3)
  {
    var triplet = (((binarray[i   >> 2] >> 8 * (3 -  i   %4)) & 0xFF) << 16)
                | (((binarray[i+1 >> 2] >> 8 * (3 - (i+1)%4)) & 0xFF) << 8 )
                |  ((binarray[i+2 >> 2] >> 8 * (3 - (i+2)%4)) & 0xFF);
    for(var j = 0; j < 4; j++)
    {
      if(i * 8 + j * 6 > binarray.length * 32) str += b64pad;
      else str += tab.charAt((triplet >> 6*(3-j)) & 0x3F);
    }
  }
  return str;
}
EOS

}	# End of javascript.

# -----------------------------------------------

sub new
{
	my($class, %arg)	= @_;
	my($self)			= bless({}, $class);

	for my $attr_name ($self -> _standard_keys() )
	{
		my($arg_name) = $attr_name =~ /^_(.*)/;

		if (exists($arg{$arg_name}) )
		{
			$$self{$attr_name} = $arg{$arg_name};
		}
		else
		{
			$$self{$attr_name} = $self -> _default_for($attr_name);
		}
	}

	$self;

}	# End of new.

# -----------------------------------------------

1;

__END__

=head1 NAME

C<Javascript::SHA1> - Calculate the SHA1 digest of a CGI form field

=head1 Synopsis

This is a complete, runnable, tested program.

	#!/usr/bin/perl

	use strict;
	use warnings;

	use CGI;
	use Javascript::SHA1;

	# ------------------

	my($q)	= CGI -> new();
	my($p)	= $q -> param('my_password') || '';
	my($js)	= Javascript::SHA1 -> new();

	print $q -> header(),
      $q -> start_html({script => $js -> javascript('my_password'), title => 'Javascript::SHA1'}),
      $q -> h1({align => 'center'}, 'Javascript::SHA1'),
      "Previous value: $p",
      $q -> br(),
      $q -> start_form({action => $q -> url(), name => 'SHA1'}),
      'Username: ',
      $q -> textfield({name => 'my_username', size => 50}),
      $q -> br(),
      'Password: ',
      $q -> password_field({name => 'my_password', size => 50}),
      $q -> br(),
      'Generate str2hex_sha1: ',
      $q -> submit({onClick => 'return str2hex_sha1()'}),
      $q -> end_form(),
      $q -> end_html();

=head1 Description

C<Javascript::SHA1> is a pure Perl module.

This module allows you to convert user input, eg a password, into the SHA1 digest of that
input.

This means the password itself need never be transmitted across the network - only the SHA1
digest need be transmitted.

=head1 Distributions

This module is available both as a Unix-style distro (*.tgz) and an
ActiveState-style distro (*.ppd). The latter is shipped in a *.zip file.

See http://savage.net.au/Perl-modules.html for details.

See http://savage.net.au/Perl-modules/html/installing-a-module.html for
help on unpacking and installing each type of distro.

=head1 Constructor and initialization

new(...) returns a C<Javascript::SHA1> object.

This is the class's contructor.

Usage: Javascript::SHA1 -> new().

C<new()> does not take any parameters.

=head1 Method: javascript([$name_of_CGI_password_field][, $number_of_CGI_form])

Returns a block of Javascript which you must output as part of your HTML page.

Takes 2 optional parameters:

=over 4

=item The name of the CGI field used to input the password

This field name defaults to 'password'.

=item The number of the CGI form within the HTML page

This field value defaults to 0, which is the first form on the page.

=back

See the Synopsis for one way to do this.

Note: if you pass just 1 parameter, it is assumed to be the name of a field.

=head1 Submitting a CGI Form

To disable this module, simply use a submit button of the form:

	$q -> submit();

To enable the module, use a submit button with an onClick event handler that calls a
Javascript function which converts your CGI password field's value into the SHA1 digest
of that value. Use one of:

=over 4

=item $q -> submit({onClick => 'return str2hex_sha1()'});

Returns a string of length 40.

=item $q -> submit({onClick => 'return str2b64_sha1()'});

Returns a string of length 27.

=item $q -> submit({onClick => 'return str2str_sha1()'});

Returns a string of length 20 (approx :-).

=back

=head1 Example code

See the examples/ directory in the distro.

There is 1 demo file, sha1.cgi.

This is a CGI script identical to the one in the synopsis.

Edit it to suit your circumstances.

=head1 Related Modules

Javascript::MD5.

=head1 Required Modules

None.

=head1 Author

C<Javascript::SHA1> was written by Ron Savage I<E<lt>ron@savage.net.auE<gt>> in 2004.

Home page: http://savage.net.au/index.html

=head1 Copyright

Australian copyright (c) 2004, Ron Savage. All rights reserved.

	All Programs of mine are 'OSI Certified Open Source Software';
	you can redistribute them and/or modify them under the terms of
	The Artistic License, a copy of which is available at:
	http://www.opensource.org/licenses/index.html

=cut
